lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp

```diff
@@ -0,0 +1,181 @@
+  /// not have a target.
+  auto findAndEraseTarget =
+      [](llvm::json::Object *object,
+         llvm::json::Path p) -> llvm::Optional<std::string> {
+    // If no "target" field exists, then promote the annotation to a
+    // CircuitTarget annotation by returning a target of "~".
+    auto target = object->get("target");
+    if (!target)
+      return llvm::Optional<std::string>("~");
+
+    // Find the target.
+    auto maybeTarget = object->get("target")->getAsString();
+
+    // If this is a normal Target (not a Named), erase that field in the JSON
+    // object and return that Target.
+    std::string newTarget;
+    if (maybeTarget.getValue()[0] == '~') {
+      newTarget = maybeTarget->str();
+    } else {
+      // This is a legacy target using the firrtl.annotations.Named type.  This
+      // can be trivially canonicalized to a non-legacy target, so we do it with
+      // the following thre mappings:
+      //   1. CircuitName => CircuitTarget, e.g., A -> ~A
+      //   2. ModuleName => ModuleTarget, e.g., A.B -> ~A|B
+      //   3. ComponentName => ReferenceTarget, e.g., A.B.C -> ~A|B>C
+      newTarget = "~";
+      llvm::raw_string_ostream s(newTarget);
+      bool isModule = true;
+      for (auto a : maybeTarget.getValue()) {
+        switch (a) {
+        case '.':
+          if (isModule) {
+            s << "|";
+            isModule = false;
+            break;
+          }
+          s << ">";
+          break;
+        default:
+          s << a;
+        }
+      }
+    }
+
+    // If the target is something that we know we don't support, then error.
+    bool unsupported =
+        std::any_of(newTarget.begin(), newTarget.end(), [](char a) {
+          return a == '/' || a == ':' || a == '>' || a == '.' || a == '[';
+        });
+    if (unsupported) {
+      p.field("target").report(
+          "Unsupported target (not a CircuitTarget or ModuleTarget)");
+      return {};
+    }
+
+    // Remove the target field from the annotation and return the target.
+    object->erase("target");
+    return llvm::Optional<std::string>(newTarget);
+  };
+
+  /// Convert arbitrary JSON to an MLIR Attribute.
+  std::function<Attribute(llvm::json::Value &, llvm::json::Path)>
+      JSONToAttribute =
+          [&](llvm::json::Value &value, llvm::json::Path p) -> Attribute {
+    // String
+    if (auto a = value.getAsString())
+      return StringAttr::get(context, a.getValue());
+
+    // Integer
+    if (auto a = value.getAsInteger())
+      return IntegerAttr::get(IntegerType::get(context, 64), a.getValue());
+
+    // Float
+    if (auto a = value.getAsNumber())
+      return FloatAttr::get(mlir::FloatType::getF64(context), a.getValue());
+
+    // Boolean
+    if (auto a = value.getAsBoolean())
+      return BoolAttr::get(context, a.getValue());
+
+    // Null
+    if (auto a = value.getAsNull()) {
+      p.report("Found a null in the JSON, returning 'nullptr'");
+      return nullptr;
```

Comment:
You could represent this as "unit" the empty tuple attr if you wanted to.  Ths would make it so anything would round trip and would fix the missing null error handling at the callers of JSONToAttribute

Reply:
Changed to `{}`.